fedimint-api-client/src/api/mod.rs

```diff
@@ -194,92 +195,155 @@ pub trait FederationApiExt: IRawFederationApi {
+
+        for peer in self.all_peers() {
+            futures.push(Box::pin({
+                let method = &method;
+                let params = &params;
+                async move {
+                    let response = util::retry(
+                        "api-request-{method}-{peer}",
+                        api_networking_backoff(),
+                        || async {
+                            self.request_single_peer_typed(
+                                None,
+                                method.clone(),
+                                params.clone(),
+                                *peer,
+                            )
+                            .await
+                            .map_err(|e| anyhow!(e.to_string()))
+                        },
+                    )
+                    .await
+                    .expect("Number of retries has no limit");
+
+                    (*peer, response)
+                }
+            }))
+        }
+
+        loop {
+            let (peer, response) = futures
+                .next()
+                .await
+                .expect("Query strategy ran out of peers to query without returning a result");
+
+            match strategy.process(peer, response) {
+                QueryStep::Retry(peers) => {
+                    for peer in peers {
+                        futures.push(Box::pin({
+                            let method = &method;
+                            let params = &params;
+                            async move {
+                                let response = util::retry(
+                                    "api-request-{method}-{peer}",
+                                    api_networking_backoff(),
+                                    || async {
+                                        self.request_single_peer_typed(
+                                            None,
+                                            method.clone(),
+                                            params.clone(),
+                                            peer,
+                                        )
+                                        .await
+                                        .map_err(|e| anyhow!(e.to_string()))
+                                    },
+                                )
+                                .await
+                                .expect("Number of retries has no limit");
+
+                                (peer, response)
+                            }
+                        }));
                     }
                 }
-                None => {
-                    panic!("Query strategy ran out of peers to query without returning a result");
+                QueryStep::Success(response) => return response,
+                QueryStep::Failure(e) => {
+                    warn!("Query strategy returned non-retryable failure for peer {peer}: {e}");
+
+                    futures.push(Box::pin(pending()));
```

Comment:
This looks wrong, no? We are queering all peers on one future here, so we just want to do nothing and let the loop continue.

Reply:
This is in fact correct so the futures unordered will never run out of futures.

Reply:
Ohhh... . I see it now. Why do we need to do that? We can clearly detect it (running out of futures) in:

```
            let (peer, response) = futures
                .next()
                .await
                .expect("Query strategy ran out of peers to query without returning a result");
```

and running out futures (no decision despite responses from all peers) used to be a programming error.